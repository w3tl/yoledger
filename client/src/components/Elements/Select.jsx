import React from 'react';
import PropTypes from 'prop-types';
import { withLabel } from './utils';

class Select extends React.Component {
  static propTypes = {
    value: PropTypes.string,
    options: PropTypes.arrayOf(PropTypes.shape({
      value: PropTypes.string,
      label: PropTypes.string,
    })).isRequired,
    disabled: PropTypes.bool,
    onChange: PropTypes.func,
  }

  static defaultProps = {
    value: '',
    disabled: false,
    onChange: () => {},
  }

  constructor(props) {
    super(props);
    this.state = { value: props.value };
  }

  handleChange = (e) => {
    const { onChange } = this.props;
    this.setState({ value: e.target.value });
    onChange({ target: { value: e.target.value } });
  }

  render() {
    const { value } = this.state;
    const { disabled, options } = this.props;

    return (
      <select value={value} onChange={this.handleChange} disabled={disabled}>
        {options.map(opt => (
          <option key={opt.value} value={opt.value}>{opt.label}</option>
        ))}
      </select>
    );
  }
}

export default withLabel(Select);
