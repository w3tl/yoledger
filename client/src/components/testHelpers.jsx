import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

export const withProvider = Component => props => (
  <MockedProvider {...props}>
    <Component />
  </MockedProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return fixedDate;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
