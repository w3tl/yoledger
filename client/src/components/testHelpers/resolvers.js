import dateResolver from './date';
import assets from '../Asset/mockData';
import { expenses, incomes } from '../Categories/mockData';
import { accounts, budgets as rawBudgets } from '../Budget/mockData';
import transactions from '../Transaction/mockData';

const accountsArray = assets.concat(expenses).concat(incomes);
const budgets = rawBudgets.map(({ date, ...fields }) => ({ date: new Date(date), ...fields }));

export default {
  Date: dateResolver,
  Query: {
    account(root, { id }) {
      if (id === '666') return { name: 'new account' };
      return accountsArray.find(a => a.id === id);
    },
    accounts(root, { type }) {
      return accountsArray.filter(a => a.type === type);
    },
    budgets() {
      return { accounts };
    },
    budget(root, { account, dateStart, dateEnd }) {
      return budgets // eslint-disable-next-line max-len
        .filter(b => (
          b.account.name === account
          && b.date.getTime() >= dateStart.getTime()
          && b.date.getTime() <= new Date(dateEnd).getTime()));
    },
    transactions() {
      return transactions.map(({ date, ...other }) => ({ date: new Date(date), ...other }));
    },
    transaction(root, { id }) {
      if (id === '666') return { name: 'new account' };
      const { date, ...other } = transactions.find(a => a.id === id);
      return { date: new Date(date), ...other };
    },
  },
  Mutation: {
    upsertBudget(root, { input: { account, date, amount } }) {
      const existingBudget = budgets
        .find(b => (
          b.account.name === account.name && b.date.getTime() === new Date(date).getTime()
        ));
      const acc = accounts.find(a => a.name === account) || {
        id: '687',
        name: account,
      };
      return {
        success: true,
        budget: {
          id: '99',
          ...existingBudget,
          account: acc,
          date,
          amount,
        },
      };
    },
    addAccount(root, { input }) {
      return {
        account: { id: '666', ...input },
      };
    },
    addTransaction(root, { input }) {
      return {
        transaction: {
          id: '666',
          ...input,
          source: accountsArray.find(a => a.name === input.source),
          destination: accountsArray.find(a => a.name === input.destination),
        },
      };
    },
    deleteTransaction(root, { id }) {
      return {
        success: true,
        id,
      };
    },
    updateTransaction(root, { id, input }) {
      return {
        transaction: {
          ...transactions.find(t => t.id === id),
          id: '888',
          ...input,
          source: accountsArray.find(a => a.name === input.source),
          destination: accountsArray.find(a => a.name === input.destination),
        },
      };
    },
    signin(root, { login }) {
      return {
        token: `{\\"userId\\":\\"${login}\\"};secret;{\\"expiresIn\\":10}`,
      };
    },
    signout() {
      return {
        success: true,
      };
    },
    changePassword() {
      return {
        token: `{\\"userId\\":\\"${'admin'}\\"};secret;{\\"expiresIn\\":10}`,
      };
    },
  },
};
