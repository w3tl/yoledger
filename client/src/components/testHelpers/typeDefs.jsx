export default `
scalar Date
type Query {
  account(id: ID!): Account
  accounts(type: AccountType!): [Account]

  budgets(dateStart: Date!, count: Int = 5): Budget
  budget(date: Date!): [Allocation]

  transactions(dateStart: Date!, dateEnd: Date, page: Int = 0, itemsPerPage: Int = 20): [Transaction]
  transaction(id: ID!): Transaction
}
type Mutation {
  addAccount(input: AddAccountInput!): AddAccountPayload
  upsertBudget(input: UpsertBudgetInput!): UpsertBudgetPayload
  addTransaction(input: AddTransactionInput!): AddTransactionPayload
  deleteTransaction(id: ID!): DeleteTransactionPayload
  updateTransaction(id: ID!, input: UpdateTransactionInput!): UpdateTransactionPayload
}

type Account {
  id: ID!
  name: String!
  balance: Float
  type: AccountType!
}

type Budget {
  periods: [Date]
  accounts: [Account]
}

type Allocation {
  id: String!
  account: Account!
  amount: Float
  balance: Float
}

enum AccountType {
  ASSET
  EXPENSE
  INCOME
  VENDOR
}

input AddAccountInput {
  name: String!
  balance: Float
  type: AccountType!
}

type AddAccountPayload {
  account: Account
}

input UpsertBudgetInput {
  account: String!
  date: Date!
  amount: Float!
}

type UpsertBudgetPayload {
  success: Boolean
  allocation: Allocation
}

type Transaction {
  id: ID!
  amount: Float!
  source: Account!
  destination: Account!
  date: Date!
}
input AddTransactionInput {
  source: String!
  destination: String!
  amount: Float!
  date: Date
}
type AddTransactionPayload {
  transaction: Transaction
}
type DeleteTransactionPayload {
  success: Boolean!
  id: ID
}
input UpdateTransactionInput {
  amount: Float
  source: String
  destination: String
  date: Date
}
type UpdateTransactionPayload {
  transaction: Transaction
}
`;
