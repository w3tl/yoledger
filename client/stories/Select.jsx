import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { Select } from '../src/components/Elements';

const options = [{
  value: 'value1', label: 'Value 1',
}, {
  value: 'value2', label: 'Value 2',
}, {
  value: 'value3', label: 'Value 3',
}];

storiesOf('Elements/Select', module)
  .add('default', () => <Select options={options} onChange={action('onChange')} />)
  .add('with label', () => <Select options={options} label="Choose one" />);
