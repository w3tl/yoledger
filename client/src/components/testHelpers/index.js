import React from 'react';
import { shallow, mount } from 'enzyme';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './typeDefs';
import resolvers from './resolvers';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

const executableSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

const cache = new InMemoryCache();
const link = new SchemaLink({ schema: executableSchema });
export const client = new ApolloClient({
  link,
  cache,
});

export const withProvider = Component => props => (
  <ApolloProvider client={client}>
    <Component client={client} />
  </ApolloProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  const date = new Date(fixedDate);
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return date;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
