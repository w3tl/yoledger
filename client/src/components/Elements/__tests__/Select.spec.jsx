import React from 'react';
import { shallow } from 'enzyme';
import Select from '../Select';
import { testRenderWithoutError } from '../../testHelpers/index';

const options = [{
  value: 'value1', label: 'Value 1',
}, {
  value: 'value2', label: 'Value 2',
}, {
  value: 'value3', label: 'Value 3',
}];

describe('Select', () => {
  testRenderWithoutError(<Select options={options} />);

  it('should handle change', () => {
    const onChange = jest.fn();
    const output = shallow(
      <Select options={options} onChange={onChange} />,
    );
    output.simulate('change', { target: { value: 'value2' } });
    expect(onChange).toHaveBeenCalledWith({ target: { value: 'value2' } });
  });
});
