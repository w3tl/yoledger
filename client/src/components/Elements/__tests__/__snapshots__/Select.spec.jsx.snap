// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders without error 1`] = `
<Select
  disabled={false}
  id=""
  onChange={[Function]}
  options={
    Array [
      Object {
        "label": "Value 1",
        "value": "value1",
      },
      Object {
        "label": "Value 2",
        "value": "value2",
      },
      Object {
        "label": "Value 3",
        "value": "value3",
      },
    ]
  }
  value=""
/>
`;
